src/Functions/array/arrayAUC.cpp

```diff
@@ -249,52 +281,107 @@ class FunctionArrayAUC : public IFunction
+                    partial_auc_offsets->getData().getUInt(partial_auc_offsets_size * i),
+                    partial_auc_offsets->getData().getUInt(partial_auc_offsets_size * i + 1),
+                    partial_auc_offsets->getData().getUInt(partial_auc_offsets_size * i + 2),
+                    partial_auc_offsets_size == 4 ? partial_auc_offsets->getData().getUInt(partial_auc_offsets_size * i + 3) : 0);
+            else
+                result[i] = apply(scores, labels, current_offset, next_offset, scale);
             current_offset = next_offset;
         }
     }
 
 public:
     String getName() const override { return name; }
 
-    bool isVariadic() const override { return PR ? false : true; }
-    size_t getNumberOfArguments() const override { return PR ? 2 : 0; }
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
 
     bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo &) const override { return true; }
 
     DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
     {
         size_t number_of_arguments = arguments.size();
 
-        if ((!PR && (number_of_arguments < 2 || number_of_arguments > 3))
-            || (PR && number_of_arguments != 2))
-            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
-                            "Number of arguments for function {} doesn't match: passed {}, should be {}",
-                            getName(), number_of_arguments, PR ? "2" : "2 or 3");
+        if ((is_pr && (number_of_arguments < 2 || number_of_arguments > 3))
+            || (!is_pr && (number_of_arguments < 2 || number_of_arguments > 4)))
+            throw Exception(
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be {}",
+                getName(),
+                number_of_arguments,
+                is_pr ? "2 or 3" : "2, 3 or 4");
 
+        /// `arr_scores` and `arr_labels` arguments validation
         for (size_t i = 0; i < 2; ++i)
         {
             const DataTypeArray * array_type = checkAndGetDataType<DataTypeArray>(arguments[i].type.get());
             if (!array_type)
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The two first arguments for function {} must be of type Array.", getName());
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The two first arguments for function {} must be of type Array", getName());
 
             const auto & nested_type = array_type->getNestedType();
-            if (!isNativeNumber(nested_type) && !isEnum(nested_type))
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "{} cannot process values of type {}", getName(), nested_type->getName());
+            /// First argument (arr_scores) must be an array of numbers
```

Comment:
Perhaps the code will be nicer if we remove the loop in l. 332 and check both arguments one after another.